### ISSUE TITLE
HLSL: Right hand side is ignored during 

### ISSUE THREAD

**Issue by tafuri** [KhronosGroup/glslang#1176]
Declaring a buffer block, for example a RWStructuredBuffer with an assignment, ignores the right hand side. Making the following code compile without errors or warnings.
The assignment should clearly validate the assigned type and also copy over any applicable layout specifications, such as `binding` and `set`.

```hlsl
static const float TestValue = 1;

void main()
{
    RWStructuredBuffer<uint> values = TestValue;
    values[0] = 0;
}
```

**Comment by johnkslang** [comment]
Couple background points:
- Vulkan does not support local buffers, they must be at global scope
- glslang is not an HLSL validator

However, it is okay to catch and report errors when the rules are understood well enough, and if post-front-end legalization eliminates the local buffer, resulting in only global buffers, this can work.

Do we know that if the assignment occurs that legalization passes eliminate it again?

**Comment by tafuri** [comment]
I may have focused on the wrong thing with this issue so lets look at another example.

```hlsl
layout(set=4, binding=5)
RWStructuredBuffer<uint> Input;

void main()
{
    RWStructuredBuffer<uint> values = Input;
    values[0] = 0;
}
```

In HLSL `values` is basically a reference to `Input`. glslang
generates code which behaves in a similar way in the case where Input is bound to (0,0). But in the example above the layout specifications on Input get lost in the assignment, without any warning, and `values` will still be bound to (0,0).

A current workaround is to specify layout on `values` also, but that seems like a less than ideal workaround.

```hlsl
layout(set=4, binding=5)
RWStructuredBuffer<uint> Input;

void main()
{
    layout(set=4, binding=5)
    RWStructuredBuffer<uint> values = Input;
    values[0] = 0;
}
```

**Comment by johnkslang** [comment]
The problem is that assignments to buffers are not yet supported.  This one gives an error message:

```
layout(set=4, binding=5)
RWStructuredBuffer<uint> Input;

void main()
{
    RWStructuredBuffer<uint> values;
    values = Input;
    values[0] = 0;
}
```

I would agree the initializer case should give the same error message, but I think it's not what you really want.  I'm guessing what you really need is:

- `values` is a local variable, not a global uniform
- the assignment/initializer to buffers works
- SPIRV-opt will copy propagate `Input` to where `values` is used, because it is illegal to assign to buffers
- glslang will generate the illegal code (assigns to buffer) and SPIRV-opt will legalize it, per the previous bullet

**Comment by tafuri** [comment]
Yes, that to me seems like a very close match to how HLSL behaves. Looking at code generated by fxc or dxc shows that there is no difference in using the `Input`or `values` variable. The `values` variable behaves effectively like an alias in this case.

**Comment by johnkslang** [comment]
Right then. So the real issue here is that we don't yet support that buffer aliasing (we do support texture/sampler aliasing though).

I can make the error reporting more complete, and then turn this into an issue for that. Any input on relative priority of this?

**Comment by johnkslang** [comment]
Error caught by 8116cfc9bf1b48f61f85c8d059879b7ef2c34efb.

**Comment by tafuri** [comment]
Hmm, no not really. Since there are workarounds for this issue, like manually specifying set and binding for all aliases, it doesn't seem to be too urgent. Not sure how difficult it would be to add real support for this though, My quite hacky implemenation seemed to work pretty well for all cases I tested.

**Comment by johnkslang** [comment]
Thanks.  Doing it fully correctly requires function parameter formal types to orthogonally represent in/out/inout as well as buffer qualifiers, which is not yet expressible.
